rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,574 @@
+message PingRequest {}
+
+// PingResponse is a response to Ping.
+message PingResponse {}
+
+// SignRequest is a request to perform a signature with a specific hardware private key.
+message SignRequest {
+  // KeyRef references a specific hardware private key.
+  KeyRef key_ref = 1;
+  // PublicKeyDer is a public key encoded in PKIX, ASN.1 DER form.
+  // This is provided to verify that the hardware private key has not
+  // been swapped out against the public key expected by the client.
+  bytes public_key_der = 2;
+  // Digest is a hashed message to sign.
+  bytes digest = 3;
+  // Hash is the hash function used to prepare the digest.
+  Hash hash = 4;
+  // SaltLength specifies the length of the salt added to the digest before a signature.
+  // Only used, and required, for PSS RSA signatures.
+  uint32 salt_length = 5;
+}
+
+// Signature is a private key signature.
+message Signature {
+  // For an (EC)DSA key, the default key algorithm for hardware private keys, this
+  // will be a DER-serialised, ASN.1 signature structure.
+  //
+  // When the client is using a manually generated RSA key, this can be either a
+  // PKCS #1 v1.5, or if the cluster is on the legacy signature algorithm suite,
+  // a PSS signature,
+  bytes signature = 1;
+}
+
+// KeyRef references a specific hardware private key.
+message KeyRef {
+  // SerialNumber is the serial number of the hardware key.
+  uint32 serial_number = 1;
+  // PivSlot is a specific PIV slot on the hardware key.
+  PIVSlot piv_slot = 2;
+}
+
+// GetInfoRequest is a response to GetInfo
+message GetInfoRequest {
+  // KeyRef references a specific hardware private key.
+  KeyRef key_ref = 1;
+}
+
+// GetInfoResponse is a response to GetInfo.
+message GetInfoResponse {
+  // PublicKey is the hardware public key encoded in PKIX, ASN.1 DER form.
+  bytes public_key_der = 1;
+  // AttestationStatement is a hardware key attestation statement corresponding
+  // to a hardware private key.
+  teleport.attestation.v1.AttestationStatement attestation_statement = 2;
+  // PinRequired indicates whether the hardware private key requires PIN.
+  bool pin_required = 3;
+  // TouchRequired indicates whether the hardware private key requires touch.
+  bool touch_required = 4;
+}
+
+// PIVSlot is a specific PIV slot on a hardware key.
+enum PIVSlot {
+  // PIV slot not specified.
+  PIV_SLOT_UNSPECIFIED = 0;
+  // PIV slot 9a. This is the default slot for pin_policy=never, touch_policy=never.
+  PIV_SLOT_9A = 1;
+  // PIV slot 9c. This is the default slot for pin_policy=never, touch_policy=cached.
+  PIV_SLOT_9C = 2;
+  // PIV slot 9d. This is the default slot for pin_policy=once, touch_policy=cached.
+  PIV_SLOT_9D = 3;
+  // PIV slot 9e. This is the default slot for pin_policy=once, touch_policy=never.
+  PIV_SLOT_9E = 4;
+}
+
+// Hash refers to a specific hash function used during signing.
+enum Hash {
+  HASH_NAME_UNSPECIFIED = 0;
+  HASH_NAME_SHA256 = 1;
+  HASH_NAME_SHA512 = 2;
+}
```

Comment:
you'll probably need a value for "no hash", ed25519 signatures do not pre-hash, they sign the whole message

Reply:
I left it out because YubiKeys don't currently support ed25519 keys, but it's probably a good idea anyways.

Reply:
i think i missed that this was only for hardware keys now

Reply:
The zero value for a protobuf enum should not be used for anything except to implicitly signal that whatever was on the other end of the message didn't know about the enum or used some enum variant that you don't know about.

Reply:
Isn't it also ok to use it as an "unset" value? Not as specifically "none", but just "client did not specify"? Looking at the proto best practices it seems that this is the grey area.

Reply:
i think the zero value is fine in the general case to signal "client did not specify" as long as we agree that "no hash" is distinct from "unspecified" and if we want to support it we should have an explicit enum value for that